### PR TITLE
feat: add `ChannelStateChange.hasBacklog` and return state change to attach promise/callback

### LINKED ISSUE #1328: add `ChannelStateChange.hasBacklog`
This flag is already send as part of the attached protocol message, but should be exposed on `ChannelStateChange` objects so that users can tell whether to expect rewind messages.

### PR DESCRIPTION
**Pull request by owenpearson**
Resolves #1328 

See [corresponding spec PR](https://github.com/ably/specification/pull/152)

These features in combination will allow users to check whether to expect a backlog of messages once attached to a realtime channel. In particular, when using `rewind=1`, this will allow users to create a promise which resolves once a channel is attached and, if applicable, the rewind message has been received.

### CODE REVIEW

**Review by SimonWoolf — COMMENTED** [review]
No objection to the idea, seems like a reasonable feature.

There's nothing ably-js specific about this; if we think it's worth doing we should have a spec PR for this so it can eventually be done in all sdks (and added to api docs). 

> which resolves once a channel is attached and, if applicable, the rewind message has been received.

not sure I understand what the last part of that sentence means -- it resolves once the channel is attached full stop, no?

**Comment by SimonWoolf** [comment]
>  my thinking was that this is unlikely to be needed in other SDKs so adding it to the spec would likely be futile Libraries may offer a ClientOptions#agents property, for use only by other Ably-authored SDKs, on a need-to-have basis.

**Comment by SimonWoolf** [comment]
:+1: 

> my thinking was that this is unlikely to be needed in other SDKs so adding it to the spec would likely be futile

Shrug, I think there's still some value in formalising the API in the spec even if we're not imminently intending to implement in other sdks. there's other bits of the spec that are not-mandatory ("Libraries may offer a ClientOptions#agents property...", "Client libraries can optionally save a round-trip request to the Ably service for expired tokens by detecting when a token has expired...", etc)

>  it wouldn't really make sense to add this promise to the library since it only works for rewind=1, any other rewind value has the possibility of a partial backlog

yeah, really for that feature you'd need realtime to tell you when the backlog is finished, and tbh I'm not sure that that'd be a good idea. (also most sensible ways of returning another promise would involve a breaking api change, and would make it a very prominent part of the api, and istm it's not worth either of those)

**Comment by lawrence-forooghian** [comment]
I wasn't aware of the rewind feature, popping a link to the [documentation](https://ably.com/docs/realtime/channels/channel-parameters/rewind#rewind-example-ably) here for future reference.

I think that exposing the `hasBacklog` property (which I can't find any documentation of, e.g. in the [protocol spec](https://sdk.ably.com/builds/ably/specification/main/protocol/) — how did you find out about it?) makes sense, and if I've understood the use case correctly then I think it's a broadly useful thing, and hence worth including in the spec.

As for the change to the signature of `attach`, I wonder whether that needs to come alongside this change. For example, a user calling `attach` has, until now, had no direct way to access the `ChannelStateChange.resumed` property, which you can imagine a user might want to use to know whether they need to fetch channel history. I _guess_ (please correct me if wrong) that our advice to users has been that to find this information out they need to listen for a channel state change using `on`, so would the same advice not apply here too? That said, if we _are_ going to change the signature of `attach` then I think it should be a spec change, yeah.

File: ably.d.ts [hunk 1]
```diff
@@ -2590,32 +2590,44 @@ declare namespace Types {
      *
      * @param event - The event name.
      * @param listener - An event listener function.
-     * @param callbackWhenAttached - A function which will be called upon completion of the channel {@link RealtimeChannelCallbacks.attach | `attach()`} operation. If the operation succeeded, then the function will be called with `null`. If it failed, the function will be called with information about the error.
+     * @param callbackWhenAttached - A function which will be called upon completion of the channel {@link RealtimeChannelCallbacks.attach | `attach()`} operation. If the operation succeeded, then the function will be called with a {@link ChannelStateChange} object. If it failed, the function will be called with information about the error.
```
Comment (lawrence-forooghian) [inline ably.d.ts:2597]: This needs to explain `null`, too.

File: ably.d.ts [hunk 2]
```diff
@@ -2694,32 +2706,32 @@ declare namespace Types {
      *
      * @param event - The event name.
      * @param listener - An event listener function.
-     * @returns A promise which resolves upon success of the channel {@link RealtimeChannelPromise.attach | `attach()`} operation and rejects with an {@link ErrorInfo} object upon its failure.
+     * @returns A promise which, upon successful attachment to the channel, will be fulfilled with a {@link ChannelStateChange} object. Upon failure, the promise will be rejected with an {@link ErrorInfo} object.
      */
-    subscribe(event: string, listener?: messageCallback<Message>): Promise<void>;
+    subscribe(event: string, listener?: messageCallback<Message>): Promise<ChannelStateChange>;
```
Comment (lawrence-forooghian) [inline ably.d.ts:2715]: Don't all these promise `subscribe` calls need to have `null` in their return type too?

File: src/common/lib/client/realtimechannel.ts [hunk 3]
```diff
@@ -296,14 +299,18 @@ class RealtimeChannel extends Channel {
        * current mode differs from requested mode */
       this._requestedFlags = _flags as API.Types.ChannelMode[];
     } else if (this.state === 'attached') {
-      callback();
+      callback(null, new ChannelStateChange('attached', 'attached'));
```
Comment (lawrence-forooghian) [inline src/common/lib/client/realtimechannel.ts:302]: Shouldn't this be `null`?

Comment (owenpearson) [reply src/common/lib/client/realtimechannel.ts]: oops, fixed

File: ably.d.ts [hunk 4]
```diff
@@ -2694,32 +2710,32 @@ declare namespace Types {
      *
      * @param event - The event name.
      * @param listener - An event listener function.
-     * @returns A promise which resolves upon success of the channel {@link RealtimeChannelPromise.attach | `attach()`} operation and rejects with an {@link ErrorInfo} object upon its failure.
+     * @returns A promise which, upon successful attachment to the channel, will be fulfilled with a {@link ChannelStateChange} object. Upon failure, the promise will be rejected with an {@link ErrorInfo} object.
```
Comment (lawrence-forooghian) [inline ably.d.ts:2713]: This one is still missing `null` explanation?

Comment (owenpearson) [reply ably.d.ts]: fixed

**Review by lawrence-forooghian — APPROVED** [review]

**Pull request merged**